File: app/src/main/java/com/supercilex/robotscouter/ui/PermissionRequestHandler.java
package com.supercilex.robotscouter.ui;

import android.support.annotation.NonNull;
import android.support.annotation.StringRes;
import android.support.v4.app.Fragment;

import com.google.android.gms.tasks.OnSuccessListener;

import java.util.Collections;
import java.util.List;

import pub.devrel.easypermissions.AppSettingsDialog;
import pub.devrel.easypermissions.EasyPermissions;

public class PermissionRequestHandler implements EasyPermissions.PermissionCallbacks {
    private static final int WRITE_RC = 8653;

    private final List<String> mPerms;
    private final Fragment mFragment;
    private final OnSuccessListener<Void> mListener;

    public PermissionRequestHandler(List<String> perms,
                                    Fragment fragment,
                                    OnSuccessListener<Void> listener) {
        mPerms = Collections.unmodifiableList(perms);
        mFragment = fragment;
        mListener = listener;
    }

    public String[] getPermsArray() {
        return mPerms.toArray(new String[mPerms.size()]);
    }

    public void onActivityResult(int requestCode) {
        if (requestCode == AppSettingsDialog.DEFAULT_SETTINGS_REQ_CODE
                && EasyPermissions.hasPermissions(mFragment.getContext(), getPermsArray())) {
            mListener.onSuccess(null);
        }
    }

    public void requestPerms(@StringRes int rationaleId) {
        EasyPermissions.requestPermissions(mFragment,
                                           mFragment.getString(rationaleId),
                                           WRITE_RC,
                                           mPerms.toArray(new String[mPerms.size()]));
    }

    @Override
    public void onPermissionsGranted(int requestCode, List<String> perms) {
        if (requestCode == WRITE_RC) mListener.onSuccess(null);
    }

    @Override
    public void onPermissionsDenied(int requestCode, List<String> perms) {
        if (requestCode == WRITE_RC
                && EasyPermissions.somePermissionPermanentlyDenied(mFragment, perms)) {
            new AppSettingsDialog.Builder(mFragment).build().show();
        }
    }

    @Override
    public void onRequestPermissionsResult(int requestCode,
                                           @NonNull String[] permissions,
                                           @NonNull int[] grantResults) {
        EasyPermissions.onRequestPermissionsResult(requestCode,
                                                   permissions,
                                                   grantResults,
                                                   this);
    }
}

File: app/src/main/java/com/supercilex/robotscouter/util/ui/PermissionRequestHandler.kt
package com.supercilex.robotscouter.util.ui

import android.arch.lifecycle.LiveData
import android.content.Intent
import android.support.annotation.StringRes
import android.support.v4.app.Fragment
import com.supercilex.robotscouter.RobotScouter
import com.supercilex.robotscouter.util.data.ViewModelBase
import pub.devrel.easypermissions.AppSettingsDialog
import pub.devrel.easypermissions.EasyPermissions

class PermissionRequestHandler : ViewModelBase<List<String>>(), OnActivityResult {
    private val _onGranted = SingleLiveEvent<List<String>>()
    val onGranted: LiveData<List<String>> get() = _onGranted

    lateinit var perms: List<String>
        private set

    override fun onCreate(args: List<String>) {
        perms = args.toList()
    }

    override fun onActivityResult(requestCode: Int, resultCode: Int, data: Intent?) {
        if (requestCode == AppSettingsDialog.DEFAULT_SETTINGS_REQ_CODE
                && EasyPermissions.hasPermissions(RobotScouter, *perms.toTypedArray())) {
            _onGranted.value = perms
        }
    }

    fun requestPerms(host: Fragment, @StringRes rationaleId: Int) {
        EasyPermissions.requestPermissions(
                host, RobotScouter.getString(rationaleId), WRITE_RC, *perms.toTypedArray())
    }

    fun onRequestPermissionsResult(
            host: Fragment,
            requestCode: Int,
            permissions: Array<String>,
            grantResults: IntArray
    ) = EasyPermissions.onRequestPermissionsResult(
            requestCode,
            permissions,
            grantResults,
            object : EasyPermissions.PermissionCallbacks {
                override fun onPermissionsGranted(requestCode: Int, perms: List<String>) {
                    if (requestCode == WRITE_RC) _onGranted.value = perms
                }

                override fun onPermissionsDenied(requestCode: Int, perms: List<String>) {
                    if (requestCode == WRITE_RC
                            && EasyPermissions.somePermissionPermanentlyDenied(host, perms)) {
                        AppSettingsDialog.Builder(host).build().show()
                    }
                }

                override fun onRequestPermissionsResult(
                        requestCode: Int,
                        permissions: Array<out String>,
                        grantResults: IntArray
                ) = Unit
            }
    )

    private companion object {
        const val WRITE_RC = 8653
    }
}
